bigchaindb/tendermint/lib.py

```diff
@@ -273,41 +273,29 @@ def get_latest_block(self):
 
         return backend.query.get_latest_block(self.connection)
 
-    def get_block(self, block_id, include_status=False):
-        """Get the block with the specified `block_id` (and optionally its status)
+    def get_block(self, block_id):
+        """Get the block with the specified `block_id`.
 
         Returns the block corresponding to `block_id` or None if no match is
         found.
 
         Args:
             block_id (str): block id of the block to get
-            include_status (bool): also return the status of the block
-                       the return value is then a tuple: (block, status)
         """
-        # get block from database
-        if isinstance(block_id, str):
-            block_id = int(block_id)
 
         block = backend.query.get_block(self.connection, block_id)
+
+        if not block and block_id > self.get_latest_block_height_from_tendermint():
+            return
+
+        result = {'height': block_id,
```

Comment:
I think, if the replay is still in progress, we can hit the following:

```
last_stored_block < block < last_tendermint_block
```

That's why we should check for the last block stored in MongoDB.

Reply:
I think we should disable the HTTP API while BigchainDB is syncing. Tendermint is not available while syncing, if you try to push a transaction the endpoint is not reachable.

Would this solve the issue?

Reply:
I think, it would not. What if a new block arrives to Tendermint just in time so that the syncing has not yet begun while the RPC endpoint returns the new height?

Also, not accepting transactions "at times" is bad user experience apparently. If Tendermint is really not available while syncing, we can accept and store transactions ourselves.

Reply:
> I think, it would not. What if a new block arrives to Tendermint just in time so that the syncing has not yet begun while the RPC endpoint returns the new height?

This seems like a valid concern, in a situation wherein 2/3 of the network has voted in new blocks, the height retrieved from Tendermint might be way into the future and fetching the block from MongoDB may return `None`.

> Also, not accepting transactions "at times" is bad user experience apparently. If Tendermint is really not available while syncing, we can accept and store transactions ourselves.
 
I don't necessarily agree with this i.e. the system shouldn't give the illusion of being in a working state when its actually not. (Tendermint follows this approach)

Reply:
> I think, it would not. What if a new block arrives to Tendermint just in time so that the syncing has not yet begun while the RPC endpoint returns the new height?

OK now I understand, good point. I guess this is what @kansi highlighted as well. The only inconsistency that can happen is that BigchainDB latest height is one block behind Tendermint latest height. This happens if and only if the last block created by Tendermint is an empty one. I think it's not an issue.

> Also, not accepting transactions "at times" is bad user experience apparently. If Tendermint is really not available while syncing, we can accept and store transactions ourselves.

_It might be that we are using sync in different ways, for me sync happens when a node was offline for a while and needs to download the missing blocks from its peers, or when a new node is added to the network._

If a node is syncing then it does not have enough information to answer correctly to client's requests, so it should not be reachable. We can move the conversation to a specific issue.

Reply:
It's not an illusion - it's an eventually consistent system. Unless you use the synchronous mode of posting transactions, you do not immediately see your transactions in the blockchain.

The fact Tendermint follows this approach does not imply the app has to do the same. On the contrary - they could have decided to do that to let the apps handle these problems.

Reply:
> This happens if and only if the last block created by Tendermint is an empty one. I think it's not an issue.

And what if the last block was created by other nodes? It's still fine to return 404 as long as this node does not have the information about it.

> If a node is syncing then it does not have enough information to answer correctly to client's requests, so it should not be reachable

At `DeliverTx` the situation may change so we never have enough information until the transaction is committed. We may offer an interface saying smth like "ok, we can accept your transactions but we __do not__ promise anything about them apart from the minimal validity; later your can check their status" or we can say "come back later" what can be convenient for us but not necessarily convenient for the clients. But I agree - it's out of scope of this PR.

Reply:
> It's not an illusion - it's an eventually consistent system

I think the discussion is going off topic but just to clarify Tendermint is a strongly consistent system with consensus guranteed at each block unlike bitcoin or ethereum. So this has nothing to do with consistency but with user experience as you pointed out earlier.

Reply:
We kept different types of consistency in mind. Tendermint blocks, unlike bitcoin blocks, can not change when committed - in this sense Tendermint is strongly consistent. But it's eventually consistent in the conventional sense of a distributed system consistency - the nodes may disagree on what is the last block of the blockchain.